### ISSUE TITLE
Please support the renameat2 systemcall

### ISSUE THREAD

**Issue by josch** [dex4er/fakechroot#60]
Currently, when I try to use `mv` from coreutils inside a Debian unstable chroot I get an `EACCESS`:

    $ fakechroot fakeroot chroot debian-unstable /bin/mv /etc/fstab /etc/blub
    /bin/mv: cannot move '/etc/fstab' to '/etc/blub': Permission denied

Using strace I found:

    [pid  7379] renameat2(AT_FDCWD, "/etc/fstab", AT_FDCWD, "/etc/blub", RENAME_NOREPLACE) = -1 EACCES (Permission denied)

And indeed it seems that fakechroot does not yet support the `renameat2` systemcall.

**Comment by guillemj** [comment]
Given that there's no wrapper provided by glibc for this syscall, you'd need to wrap syscall(2) itself, handle this specific SYS_renameat2 and otherwise pass control to the original syscall(2) via __builtin_apply_args + __builtin_apply + __builtin_return at least with gcc. Although computing the size argument seems non-obvious.

This of course will not cover the case of code using the _syscall(2) macro, as that gets inlined in the code.

**Comment by josch** [comment]
In case of coreutils `mv` it seems to be using pain `syscall(2)`: https://sources.debian.org/src/coreutils/8.30-1/lib/renameat2.c/?hl=81#L81

**Comment by josch** [comment]
Here is a patch that adds support for the renameat2 glibc function:

```patch
--- fakechroot-2.19.orig/config.h.in
+++ fakechroot-2.19/config.h.in
@@ -526,6 +526,9 @@
 /* Define to 1 if you have the `renameat' function. */
 #undef HAVE_RENAMEAT
 
+/* Define to 1 if you have the `renameat2' function. */
+#undef HAVE_RENAMEAT2
+
 /* Define to 1 if you have the `revoke' function. */
 #undef HAVE_REVOKE
 
--- fakechroot-2.19.orig/configure.ac
+++ fakechroot-2.19/configure.ac
@@ -252,6 +252,7 @@ AC_CHECK_FUNCS(m4_normalize([
     removexattr
     rename
     renameat
+    renameat2
     revoke
     rmdir
     scandir
--- fakechroot-2.19.orig/src/Makefile.am
+++ fakechroot-2.19/src/Makefile.am
@@ -135,6 +135,7 @@ libfakechroot_la_SOURCES = \
     removexattr.c \
     rename.c \
     renameat.c \
+    renameat2.c \
     revoke.c \
     rmdir.c \
     rpl_lstat.c \
--- /dev/null
+++ fakechroot-2.19/src/renameat2.c
@@ -0,0 +1,42 @@
+/*
+    libfakechroot -- fake chroot environment
+    Copyright (c) 2010, 2013 Piotr Roszatycki <dexter@debian.org>
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2.1 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
+*/
+
+
+#include <config.h>
+
+#ifdef HAVE_RENAMEAT2
+
+#define _ATFILE_SOURCE
+#include "libfakechroot.h"
+
+
+wrapper(renameat2, int, (int olddirfd, const char * oldpath, int newdirfd, const char * newpath, unsigned int flags))
+{
+    char tmp[FAKECHROOT_PATH_MAX];
+    debug("renameat2(%d, \"%s\", %d, \"%s\", %d)", olddirfd, oldpath, newdirfd, newpath, flags);
+    expand_chroot_path_at(olddirfd, oldpath);
+    strcpy(tmp, oldpath);
+    oldpath = tmp;
+    expand_chroot_path_at(newdirfd, newpath);
+    return nextcall(renameat2)(olddirfd, oldpath, newdirfd, newpath, flags);
+}
+
+#else
+typedef int empty_translation_unit;
+#endif
```

**Comment by dex4er** [comment]
Fixed in 3ecd87aa5e851b70cea0a6991720ccb6ce28106f
